platform/security/idp/security-idp-server/src/main/java/org/codice/ddf/security/idp/server/IdpEndpoint.java

```diff
@@ -0,0 +1,820 @@
+                    password = password.trim();
+                    password = decryptPassword(password, encryptionService);
+                }
+                return password;
+            }
+        }
+        return null;
+    }
+
+    protected String decryptPassword(String password, EncryptionService passwordEncryptor) {
+        if (password.startsWith(Merlin.ENCRYPTED_PASSWORD_PREFIX) && password
+                .endsWith(Merlin.ENCRYPTED_PASSWORD_SUFFIX)) {
+            if (passwordEncryptor == null) {
+                LOGGER.debug("No PasswordEncryptor is configured!");
+                return password;
+            }
+            String substring = password
+                    .substring(Merlin.ENCRYPTED_PASSWORD_PREFIX.length(), password.length() - 1);
+            return passwordEncryptor.decrypt(substring);
+        }
+
+        return password;
+    }
+
+    private static class AuthnContextClassTransformer implements Transformer {
+
+        @Override
+        public Object transform(Object o) {
+            if (o instanceof AuthnContextClassRef) {
+                return ((AuthnContextClassRef) o).getAuthnContextClassRef();
+            }
+            return o;
+        }
+    }
+
+    @GET
+    @Path("/metadata")
+    @Produces("application/xml")
+    public Response retrieveMetadata() throws WSSecurityException, CertificateEncodingException {
+        //TODO get this the right way
+        String servletContext = System.getProperty("org.codice.ddf.system.rootContext", "/services");
+        String hostname = System.getProperty("org.codice.ddf.system.hostname", "localhost");
+        String port = System.getProperty("org.codice.ddf.system.httpsPort", "8993");
+        List<String> nameIdFormats = new ArrayList<>();
+        nameIdFormats.add(SAML2Constants.NAMEID_FORMAT_PERSISTENT);
+        nameIdFormats.add(SAML2Constants.NAMEID_FORMAT_UNSPECIFIED);
+        nameIdFormats.add(SAML2Constants.NAMEID_FORMAT_X509_SUBJECT_NAME);
+        CryptoType cryptoType = new CryptoType(CryptoType.TYPE.ALIAS);
+        cryptoType.setAlias(getSignatureCrypto().getDefaultX509Identifier());
+        X509Certificate[] certs = getSignatureCrypto().getX509Certificates(cryptoType);
+        X509Certificate issuerCert = certs[0];
+
+        cryptoType = new CryptoType(CryptoType.TYPE.ALIAS);
+        cryptoType.setAlias(getEncryptionCrypto().getDefaultX509Identifier());
+        certs = getEncryptionCrypto().getX509Certificates(cryptoType);
+        X509Certificate encryptionCert = certs[0];
+        EntityDescriptor entityDescriptor = SamlProtocol
+                .createIdpMetadata("https://" + hostname + ":" + port + servletContext + "/idp/login",
+                        Base64.encodeBase64String(issuerCert.getEncoded()),
+                        Base64.encodeBase64String(encryptionCert.getEncoded()), nameIdFormats,
+                        "https://" + hostname + ":" + port + servletContext + "/idp/login",
```

Comment:
When refactoring use the SystemBaseUrl constructUrl("/ldp/login", true) method. You can then get rid of the System.getProperty calls above.